v2/src/log_stacktrace.c

```diff
@@ -0,0 +1,279 @@
+                stackFrame.AddrFrame.Mode = AddrModeFlat;
+                stackFrame.AddrStack.Offset = context.STACK_POINTER_REGISTER;
+                stackFrame.AddrStack.Mode = AddrModeFlat;
+
+                bool thisIsFirstFrame = true; /*used to skip the first frame, which is "us", that is don't display information about "getThreadStackAsString"*/
+
+                /*4) once the context has been acquired, call StackWalk64 to get the stack frames. For every frame:*/
+                while (StackWalk64(
+                    IMAGE_FILE_MACHINE_AMD64,
+                    hProcess,
+                    hThread,
+                    &stackFrame,
+                    &context,
+                    NULL,
+                    SymFunctionTableAccess64,
+                    SymGetModuleBase64,
+                    NULL))
+                {
+
+                    if (thisIsFirstFrame)
+                    {
+                        thisIsFirstFrame = false; /*no printing for the top of the stack, which is "us". us = getThreadStackAsString*/
+                        continue;
+                    }
+
+                    DWORD64 displacement = 0;
+
+                    SYMBOL_INFO_EXTENDED buffer;
+
+                    PSYMBOL_INFO pSymbol = &buffer.symbol;
+
+                    pSymbol->SizeOfStruct = sizeof(SYMBOL_INFO);
+                    pSymbol->MaxNameLen = MAX_SYM_NAME;
+
+                    const char* function_name;
+
+                    /*4.1) determine the function name*/
+                    if (!SymFromAddr(hProcess, stackFrame.AddrPC.Offset, &displacement, pSymbol))
+                    {
+                        function_name = "failure in SymFromAddr";
+                    }
+                    else
+                    {
+                        function_name = pSymbol->Name;
+                    }
+
+                    IMAGEHLP_LINE64 line;
+                    line.SizeOfStruct = sizeof(IMAGEHLP_LINE64);
+                    DWORD lineDisplacement;
+                    const char* file_name;
+                    uint32_t line_number;
+
+                    /*4.2) determine the file name and line number*/
+                    if (!SymGetLineFromAddr64(hProcess, stackFrame.AddrPC.Offset, &lineDisplacement, &line))
+                    {
+                        file_name = "failure in SymGetLineFromAddr64";
+                        line_number = 0;
+                    }
+                    else
+                    {
+                        file_name = line.FileName;
+                        line_number = line.LineNumber;
+                    }
+
+                    /*4.3) append the function name, file name and line number to the destination*/
+                    snprintf_fallback(&destination, &destinationSize, snprintfFailed, sizeof(snprintfFailed), "%s!%s %s:%" PRIu32 "", firstLine ? (firstLine = false, "") : "\n", function_name, file_name, line_number);
```

Comment:
>snprintf_fallback [](http://example.com/codeflow?start=20&length=17)

This might be an interesting thing to extract somehow. Because it would be useful in other parts of the SW (I am thinking when building a log line - similar logic is implemented there and maybe it would not be bad to had a snprintf_with_fallback as a separate unit.

/task is fine.

Reply:
Made a task.

Reply:
[https://msazure.visualstudio.com/One/_workitems/edit/25732754](https://msazure.visualstudio.com/One/_workitems/edit/25732754)